src/Gui/ImageView.cpp

```diff
@@ -0,0 +1,369 @@
+    reader.setAutoTransform(true);
+    QImage image = reader.read();
+    if (image.isNull()) {
+        QMessageBox::information(this, tr("Failed to load image file"),
+                                 tr("Cannot load file %1: %2")
+                                 .arg(fileName, reader.errorString()));
+        return false;
+    }
+
+    setImage(image);
+    setWindowFilePath(fileName);
+
+    return true;
+}
+
+void ImageView::setImage(const QImage& image)
+{
+    rawImage = image;
+    imageLabel->setPixmap(QPixmap::fromImage(image));
+    imageLabel->adjustSize();
+    scrollArea->setVisible(true);
+    scaleFactor = 1.0;
+}
+
+void ImageView::scaleImage(double factor)
+{
+    scaleFactor *= factor;
+#if QT_VERSION >= QT_VERSION_CHECK(5,15,0)
+    imageLabel->resize(scaleFactor * imageLabel->pixmap(Qt::ReturnByValue).size());
+#else
+    imageLabel->resize(scaleFactor * imageLabel->pixmap()->size());
+#endif
+
+    adjustScrollBar(scrollArea->horizontalScrollBar(), factor);
+    adjustScrollBar(scrollArea->verticalScrollBar(), factor);
+}
+
+void ImageView::adjustScrollBar(QScrollBar *scrollBar, double factor)
+{
+    scrollBar->setValue(int(factor * scrollBar->value()
+                            + ((factor - 1) * scrollBar->pageStep()/2)));
+}
+
+bool ImageView::canZoomIn() const
+{
+    return !isFitToWindow() && imageLabel->width() < 10000;
+}
+
+bool ImageView::canZoomOut() const
+{
+    return !isFitToWindow() && imageLabel->width() > 200;
+}
+
+void ImageView::zoomIn()
+{
+    scaleImage(1.25);
+}
+
+void ImageView::zoomOut()
+{
+    scaleImage(0.8);
+}
+
+void ImageView::normalSize()
+{
+    imageLabel->adjustSize();
+    scaleFactor = 1.0;
+}
+
+void ImageView::fitToWindow(bool fitView)
+{
+    scrollArea->setWidgetResizable(fitView);
+    if (!fitView) {
+        normalSize();
+    }
+}
+
+bool ImageView::isFitToWindow() const
+{
+    return scrollArea->widgetResizable();
+}
+
+bool ImageView::canDrag() const
+{
+    return scrollArea->verticalScrollBar()->isVisible() ||
+           scrollArea->horizontalScrollBar()->isVisible();
+}
+
+void ImageView::startDrag()
+{
+    dragging = true;
+}
+
+void ImageView::stopDrag()
+{
+    dragging = false;
+}
+
+bool ImageView::isDragging() const
+{
+    return dragging;
+}
+
+void ImageView::contextMenuEvent(QContextMenuEvent* event)
+{
+    QMenu menu;
+    QAction* fitToWindowAct = menu.addAction(tr("Fit to window"));
+    fitToWindowAct->setCheckable(true);
+    fitToWindowAct->setChecked(isFitToWindow());
+    connect(fitToWindowAct, &QAction::toggled, this, &ImageView::fitToWindow);
+
+    QAction* zoomInAct = menu.addAction(tr("Zoom in"), this, &ImageView::zoomIn);
+    zoomInAct->setEnabled(canZoomIn());
+
+    QAction* zoomOutAct = menu.addAction(tr("Zoom out"), this, &ImageView::zoomOut);
+    zoomOutAct->setEnabled(canZoomOut());
+
+    menu.exec(event->globalPos());
+}
+
+void ImageView::mousePressEvent(QMouseEvent* event)
+{
+    if (event->buttons() & Qt::MiddleButton) {
```

Comment:
Technically this is an implicit conversion from an int to a bool.

Reply:
Switched to QFlags::testFlag